src/probability/martingale/borel_cantelli.lean

```diff
@@ -0,0 +1,748 @@
+    { simp only [set.indicator_of_not_mem h, nat.cast_zero] } }
+end
+
+end borel_cantelli
+
+open borel_cantelli
+
+lemma bdd_above_range_sum_indicator_iff
+  (μ : measure Ω) [is_finite_measure μ] {s : ℕ → set Ω} (hs : ∀ n, measurable_set[ℱ n] (s n)) :
+  ∀ᵐ ω ∂μ,
+    bdd_above (set.range $ (λ n, ∑ k in finset.range n, (s (k + 1)).indicator (1 : Ω → ℝ) ω)) ↔
+    bdd_above (set.range $
+      (λ n, ∑ k in finset.range n, μ[(s (k + 1)).indicator (1 : Ω → ℝ) | ℱ k] ω)) :=
+begin
+  have h₁ := (martingale_mgale μ hs).ae_not_tendsto_at_top_at_top (mgale_diff_le' hs),
+  have h₂ := (martingale_mgale μ hs).ae_not_tendsto_at_top_at_bot (mgale_diff_le' hs),
+  have h₃ : ∀ᵐ ω ∂μ, ∀ k, (0 : ℝ) ≤ μ[(s (k + 1)).indicator 1|ℱ k] ω,
+  { rw ae_all_iff,
+    exact λ n, condexp_nonneg (eventually_of_forall $ set.indicator_nonneg $ λ _ _, zero_le_one) },
+  filter_upwards [h₁, h₂, h₃] with ω hω₁ hω₂ hω₃,
+  split; rintro ⟨b, hbdd⟩; by_contra hcon,
+  { have ht : tendsto (λ n, ∑ k in finset.range n, μ[(s (k + 1)).indicator 1|ℱ k] ω) at_top at_top,
+    { simp only [not_bdd_above_iff, set.mem_range, exists_prop, exists_exists_eq_and] at hcon,
+      have : ∀ R : ℝ, ∃ n,
+        R ≤ ∑ k in finset.range n, μ[(s (k + 1)).indicator 1|ℱ k] ω :=
+        λ R, let ⟨n, hn⟩ := hcon R in ⟨n, hn.le⟩,
+      refine (monotone_nat_of_le_succ (λ n, _)).tendsto_at_top_at_top this,
+      rw finset.sum_range_succ,
+      exact le_add_of_nonneg_right (hω₃ _) },
+    rw ← tendsto_neg_at_bot_iff at ht,
+    simp_rw [mgale, finset.sum_apply, pi.sub_apply, finset.sum_sub_distrib, sub_eq_add_neg] at hω₂,
+    exact hω₂ (tendsto_at_bot_add_left_of_ge _ b (λ n, hbdd ⟨n, rfl⟩) ht) },
+  { have ht : tendsto (λ n, ∑ k in finset.range n, (s (k + 1)).indicator 1 ω) at_top at_top,
+    { simp only [not_bdd_above_iff, set.mem_range, exists_prop, exists_exists_eq_and] at hcon,
+      have : ∀ R : ℝ, ∃ n,
+        R ≤ ∑ k in finset.range n, (s (k + 1)).indicator 1 ω :=
+        λ R, let ⟨n, hn⟩ := hcon R in ⟨n, hn.le⟩,
+      refine (monotone_nat_of_le_succ (λ n, _)).tendsto_at_top_at_top this,
+      rw finset.sum_range_succ,
+      exact le_add_of_nonneg_right (set.indicator_nonneg (λ _ _, zero_le_one) _) },
+    simp_rw [mgale, finset.sum_apply, pi.sub_apply, finset.sum_sub_distrib, sub_eq_add_neg] at hω₁,
+    exact hω₁ (tendsto_at_top_add_right_of_le _ (-b) ht $ λ n, neg_le_neg (hbdd ⟨n, rfl⟩)) },
+end
+
+/-- **Lévy's generalization of the Borel-Cantelli lemma**: given a sequence of sets `s` and a
+filtration `ℱ` such that for all `n`, `s n` is `ℱ n`-measurable, `at_top.limsup s` is almost
+everywhere equal to the set for which `∑ k, ℙ(s (k + 1) | ℱ k) = ∞`. -/
+theorem ae_mem_limsup_at_top_iff_bdd_above_range [is_finite_measure μ]
+  {s : ℕ → set Ω} (hs : ∀ n, measurable_set[ℱ n] (s n)) :
+  ∀ᵐ ω ∂μ, ω ∈ at_top.limsup s ↔
+  ¬ bdd_above (set.range $
+    (λ n, ∑ k in finset.range n, μ[(s (k + 1)).indicator (1 : Ω → ℝ) | ℱ k] ω)) :=
```

Comment:
In this statement and many other ones, you formulate things in terms of `¬ bdd_above`, but wouldn't it be more natural and easier to use to say that the sequence of sums tends to infinity at infinity?